### PR TITLE
Scroll inertia cancel for iPadOS

### PR DESCRIPTION
**Pull request by moffatman**
Send a `PointerScrollInertiaCancel` event when the user touches the trackpad. Done in two ways, as one only works on real iPad devices, and the other only works when in emulation on an Apple Silicon Mac. 

Part of https://github.com/flutter/flutter/issues/106979

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt. See [testing the engine] for instructions on
writing and running engine tests.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.

### CODE REVIEW

**Comment by jmagman** [comment]
@hellohuanlin would you mind taking a look?

File: shell/platform/darwin/ios/framework/Source/FlutterViewController.mm [hunk 1]
```diff
@@ -127,6 +127,12 @@ @implementation FlutterViewController {
   fml::scoped_nsobject<UIScrollView> _scrollView;
   fml::scoped_nsobject<UIView> _keyboardAnimationView;
   MouseState _mouseState;
+  // Timestamp after which a scroll inertia cancel event should be inferred.
+  NSTimeInterval _scrollInertiaEventStartline;
+  // When an iOS app is running in emulation on an Apple Silicon Mac, direct trackpad input is
```
Comment (hellohuanlin) [inline shell/platform/darwin/ios/framework/Source/FlutterViewController.mm:132]: what is "direct" trackpad input? in apple's terminology, "direct" input means input on the screen, so trackpad should be "indirect".

File: shell/platform/darwin/ios/framework/Source/FlutterViewController.mm [hunk 2]
```diff
@@ -1847,9 +1853,29 @@ - (BOOL)gestureRecognizer:(UIGestureRecognizer*)gestureRecognizer
   return YES;
 }
 
+- (BOOL)gestureRecognizer:(UIGestureRecognizer*)gestureRecognizer
+       shouldReceiveEvent:(UIEvent*)event API_AVAILABLE(ios(13.4)) {
+  if (gestureRecognizer == _continuousScrollingPanGestureRecognizer &&
+      event.type == UIEventTypeScroll) {
```
Comment (hellohuanlin) [inline shell/platform/darwin/ios/framework/Source/FlutterViewController.mm:1859]: do we need this `UIEventTypeScroll` check? What other event type is possible?

File: shell/platform/darwin/ios/framework/Source/FlutterViewController.mm [hunk 3]
```diff
@@ -1847,9 +1853,29 @@ - (BOOL)gestureRecognizer:(UIGestureRecognizer*)gestureRecognizer
...
+    flutter::PointerData pointer_data = [self generatePointerDataAtLastMouseLocation];
+    pointer_data.device = reinterpret_cast<int64_t>(_continuousScrollingPanGestureRecognizer);
+    pointer_data.kind = flutter::PointerData::DeviceKind::kTrackpad;
+    pointer_data.signal_kind = flutter::PointerData::SignalKind::kScrollInertiaCancel;
+
+    if (event.timestamp < _scrollInertiaEventAppKitDeadline) {
```
Comment (hellohuanlin) [inline shell/platform/darwin/ios/framework/Source/FlutterViewController.mm:1865]: can you comment out that if scroll happens before the "app kit deadline", it's likely that the cancel event is caused by user and not by the system.

File: shell/platform/darwin/ios/framework/Source/FlutterViewController.mm [hunk 4]
```diff
@@ -1874,9 +1900,33 @@ - (void)hoverEvent:(UIPanGestureRecognizer*)recognizer API_AVAILABLE(ios(13.4))
...
-  auto packet = std::make_unique<flutter::PointerDataPacket>(1);
-  packet->SetPointerData(/*index=*/0, pointer_data);
-  [_engine.get() dispatchPointerDataPacket:std::move(packet)];
+  NSTimeInterval time = [NSProcessInfo processInfo].systemUptime;
+  BOOL isRunningOnMac = NO;
+  if (@available(iOS 14.0, *)) {
+    // This "stationary pointer" heuristic is not needed when running within macOS.
+    // We instead receive a scroll cancel event directly translated from AppKit.
+    // See gestureRecognizer:shouldReceiveEvent:
+    isRunningOnMac = [NSProcessInfo processInfo].iOSAppOnMac;
```
Comment (hellohuanlin) [inline shell/platform/darwin/ios/framework/Source/FlutterViewController.mm:1909]: why are we checking `isRunningOnMac` for `hoverEvent:`, but not in  `gestureRecognizer:shouldReceiveEvent:`

File: shell/platform/darwin/ios/framework/Source/FlutterViewController.mm [hunk 5]
```diff
@@ -1929,6 +1979,21 @@ - (void)continuousScrollEvent:(UIPanGestureRecognizer*)recognizer API_AVAILABLE(
...
+      // is sent whether the momentum ended normally or was cancelled by a trackpad touch.
+      // Since Flutter scrolling inertia will likely not match the system inertia, we should
+      // only send a PointerScrollInertiaCancel event for user-initiated cancellations.
+      // The following (curve-fitted) calculation provides a cutoff point after which any
+      // UIEventTypeScroll event will likely be from the system instead of the user.
+      _scrollInertiaEventAppKitDeadline =
+          [[NSProcessInfo processInfo] systemUptime] +
+          (0.1725 * log(hypotf([recognizer velocityInView:self.view].x,
+                               [recognizer velocityInView:self.view].y))) -
+          0.1196;
```
Comment (hellohuanlin) [inline shell/platform/darwin/ios/framework/Source/FlutterViewController.mm:1996]: Where is this formula and the magic numbers coming from? Can you add link to it?

**Review by hellohuanlin — COMMENTED** [review]
Thanks for the contribution! This is not a complete review since I'm still trying to understand the logic here. Left a few questions.

File: shell/platform/darwin/ios/framework/Source/FlutterViewController.mm (hunk 4 above)
Comment (moffatman) [reply shell/platform/darwin/ios/framework/Source/FlutterViewController.mm:1909]: shouldReceiveEvent is only called when running on mac. It's never called on iPad so I didn't put a check.

File: shell/platform/darwin/ios/framework/Source/FlutterViewController.mm (hunk 2 above)
Comment (moffatman) [reply shell/platform/darwin/ios/framework/Source/FlutterViewController.mm:1859]: UIEventTypeTouches will be processed here, we need to make sure it is this scroll event that follows after use of the continuous pan gesture recognizer.

File: shell/platform/darwin/ios/framework/Source/FlutterViewController.mm (hunk 1 above)
Comment (moffatman) [reply shell/platform/darwin/ios/framework/Source/FlutterViewController.mm:132]: Direct = direct connection from trackpad to UIKit. When `runningOnMac`, it goes through AppKit first. I can reword it to avoid confusion.

File: shell/platform/darwin/ios/framework/Source/FlutterViewController.mm (hunk 5 above)
Comment (moffatman) [reply shell/platform/darwin/ios/framework/Source/FlutterViewController.mm:1996]: I collected some data myself on behavior while `runningOnMac`. Not sure where to put it, maybe in this PR comments? And I will link to this PR?

File: shell/platform/darwin/ios/framework/Source/FlutterViewController.mm (hunk 3 above)
Comment (moffatman) [reply shell/platform/darwin/ios/framework/Source/FlutterViewController.mm:1865]: Will change it :+1:

File: shell/platform/darwin/ios/framework/Source/FlutterViewController.mm (hunk 4 above)
Comment (hellohuanlin) [reply shell/platform/darwin/ios/framework/Source/FlutterViewController.mm:1909]: Interesting, any idea why shouldReceiveEvent is only called on mac? Could you also add a comment about it?

File: shell/platform/darwin/ios/framework/Source/FlutterViewController.mm (hunk 5 above)
Comment (hellohuanlin) [reply shell/platform/darwin/ios/framework/Source/FlutterViewController.mm:1996]: I was wondering where did the formula `a * log (hypot(x, y)) + b` come from. 

The `log` part is particularly interesting, because scrolling inertia is typically achieved by a constant acceleration in the opposite direction, in which case the time should be linear to delta of speed (something like `dT = (v1 - v2) / a` where `dT` is delta time, `v1` is the current speed (aka `hypotf(x, y)`), `v2` is the future speed (zero), `a` is acceleration constant. I am not sure why there is a `log` or `exponential` in the formula. (high school was ages ago so i can be terribly wrong here in physics). 

Another question is how were you able to get such precise magic numbers like `0.1725` and `0.1196`.

File: shell/platform/darwin/ios/framework/Source/FlutterViewController.mm (hunk 4 above)
Comment (moffatman) [reply shell/platform/darwin/ios/framework/Source/FlutterViewController.mm:1909]: It is called on iOS as well but only for touches. Only on mac it has `UIEventTypeScroll`, I talked to the implementor at Apple, he said that appkit `NSEvent` is translated to `UIEvent`. On iOS, those events were never implemented in the first place.

File: shell/platform/darwin/ios/framework/Source/FlutterViewController.mm (hunk 5 above)
Comment (moffatman) [reply shell/platform/darwin/ios/framework/Source/FlutterViewController.mm:1996]: It is not exact, if you see my [other PR](https://github.com/flutter/flutter/pull/108298), where I attempt to match the scrolling on Mac, the stopping time cannot be computed analytically very easily. I had to use numeric method instead. This is a close approximation line of fit computed in Excel.

File: shell/platform/darwin/ios/framework/Source/FlutterViewController.mm (hunk 1 above)
Comment (hellohuanlin) [reply shell/platform/darwin/ios/framework/Source/FlutterViewController.mm:132]: yes please remove it since `.direct` is an input type that apple already uses.

File: shell/platform/darwin/ios/framework/Source/FlutterViewController.mm (hunk 4 above)
Comment (hellohuanlin) [reply shell/platform/darwin/ios/framework/Source/FlutterViewController.mm:1909]: Doesn't iPad keyboard also support UIEventTypeScroll?

File: shell/platform/darwin/ios/framework/Source/FlutterViewController.mm (hunk 5 above)
Comment (hellohuanlin) [reply shell/platform/darwin/ios/framework/Source/FlutterViewController.mm:1996]: Very interesting to see the non-linear change of velocity. Yes please add link to that PR with the graph, otherwise it's hard to tell where that come from.

File: shell/platform/darwin/ios/framework/Source/FlutterViewController.mm (hunk 4 above)
Comment (hellohuanlin) [reply shell/platform/darwin/ios/framework/Source/FlutterViewController.mm:1909]: i'm a bit confused here. Could you add comments to beginning of both functions explaining when they will be called?

**Comment by jmagman** [comment]
@moffatman thanks for the contribution, looks like there's still a few minor updates needed (mostly comments and clarifications) to get this through.

**Comment by moffatman** [comment]
> @moffatman thanks for the contribution, looks like there's still a few minor updates needed (mostly comments and clarifications) to get this through.

Yep, will be addressing it soon. I was looking into a way to avoid having to guess with magic numbers at all, but it hasn't panned out.

**Comment by moffatman** [comment]
@hellohuanlin 

image

I updated the numbers after some further data collection. I took the excel line-of-best-fit and subtracted an additional 200ms to account for all outlier data points (we want to bias towards reducing "false positives"/unwanted scroll-cancel events). I think the reason that the numbers are not a direct relationship is because I am using the UIGestureRecognizer velocity, whereas a proper UIScrollView would have a different velocity tracking formula. 

This is quite an edge case (flutter-based app running on iPad emulation in a Mac), so this is best-effort basis.

**Comment by hellohuanlin** [comment]
@moffatman what's the relation between this PR and https://github.com/flutter/flutter/pull/108298? Is this PR ready for another review? And do you need my review on the other PR?

**Review by hellohuanlin — APPROVED** [review]
Much easier to understand with the comments

File: shell/platform/darwin/ios/framework/Source/FlutterViewController.mm [hunk 6]
```diff
@@ -114,6 +114,15 @@ @implementation FlutterViewController {
   fml::scoped_nsobject<UIScrollView> _scrollView;
   fml::scoped_nsobject<UIView> _keyboardAnimationView;
   MouseState _mouseState;
+  // Timestamp after which a scroll inertia cancel event should be inferred.
+  NSTimeInterval _scrollInertiaEventStartline;
+  // When an iOS app is running in emulation on an Apple Silicon Mac, trackpad input goes through
```
[the remark this answers is not in the document]
Comment (moffatman) [reply shell/platform/darwin/ios/framework/Source/FlutterViewController.mm:119]: It's not catalyst, since the app isn't recompiled. I don't think there is an official name for this runtime, it's exposed by this property: https://developer.apple.com/documentation/foundation/nsprocessinfo/3608556-iosapponmac

File: shell/platform/darwin/ios/framework/Source/FlutterViewController.mm [hunk 7]
```diff
@@ -114,6 +114,15 @@ @implementation FlutterViewController {
...
+  // a translation layer, and events are not received with precise deltas. Due to this, we can't
```
[the remark this answers is not in the document]
Comment (moffatman) [reply shell/platform/darwin/ios/framework/Source/FlutterViewController.mm:120]: When running unmodified iOS app on Mac, events which would usually be `CGEvent` or `NSEvent` are turned into `UIEvent`. Since the iOS code can't be modified, there is some layer (which I believe is called `UINSMouseEventTranslator` based on debug data), that acts as a bridge between UIKit and AppKit. This is an apple private implementation detail and not documented.

**Comment by chinmaygarde** [comment]
Can we land this please? Looks like all presubs are green.
